apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletTransactionHelper.java

```diff
@@ -137,6 +139,44 @@ private void startCaptureBody(Transaction transaction, String method, @Nullable
         }
     }
 
+    @Nullable
+    public static String getUserFromPrincipal(@Nullable Principal principal) {
+        if (principal == null) {
+            return null;
+        }
+
+        String userName;
+        if (principal instanceof Map) {
+            // Microsoft/Azure SSO fallback
+            // https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens
+
+            Map<?, ?> map = ((Map<?, ?>) principal);
+
+            userName = getFirstClaim(map, "preferred_username");
+            if (userName == null) {
+                userName = getFirstClaim(map, "name");
+            }
+        } else {
+            userName = principal.getName();
+        }
+
+        return userName;
+    }
+
+    @Nullable
+    private static String getFirstClaim(Map<?, ?> map, String key) {
+        // entries are stored in nested collection, even when there is a single entry
+        // https://docs.microsoft.com/en-us/azure/app-service/configure-language-java?pivots=platform-linux#tomcat-1
+        Object entry = map.get(key);
+        if (entry instanceof List) {
```

Comment:
[questions] 
If it is not a list we don't want it? 
Is it never possible to be something else?
Can the obtained list be empty, making `get(0)` throw `IndexOutOfBoundsException`?

Reply:
good point on checking the list size, I'll add that.
The only information I have on those collections comes from experimentation, the code samples refer to those as generic `Collection` but they are actually `ArrayList` so we can avoid creating an iterator, I have not seen any entry which is not a list, so I think we can revisit this later if that proves to be an issue.